### ISSUE TITLE
Memoize Submissions

### ISSUE THREAD

**Issue by sunjay** [sunjay/academy#43]
Since many of the lessons will result in users typing the exact same solutions, there is no need to re-evaluate code that has already been evaluated. If a hash of the user's concatenated submission files is the same as a previous or a pending submission, do not submit again. Use that submission's result and return that immediately.

Note that this approach can work whether the user's submission results in success or failure. In either case, we can return results faster thanks to this approach.

This also protects against users submitting the same thing over and over again. It may be possible in Haskell to compute whether programs are equivalent (i.e. Identifier changed or equivalent function was used). Need to look into that. 

- [ ] To figure out if this is actually necessary and worthwhile implementing, the first part of implementing this should be to track metrics about each submission. 
    * Basically, we should concatenate the files from all submissions (trimming extra white space), hash the result and track how many duplicate hashes we get per day/per week. We want these memoized submissions to eventually expire since we don't want to keep submissions that are never used again. When concatenating files, make sure to include filenames as separators between files so the hashes are truly unique
- [ ] Implement a LRU cache that drops more than some N (~1000) number of submissions that were recently hashed
    * This way if a memoized submission is used often it will remain cached
- [ ] TODO more checklist items for implementing this once we have some idea if it is needed (change this from Type: question once we are confirmed)
